compiler/src/vm/mod.rs

```diff
@@ -1,8 +1,854 @@
+            }
+            Channel::External => {
+                self.push_external_operation(channel, Operation::Receive { performing_fiber });
+            }
+            Channel::Nursery { .. } => unreachable!("nurseries are only sent stuff"),
+        }
+    }
+
+    fn push_external_operation(&mut self, channel: ChannelId, operation: Operation) {
+        self.external_operations
+            .entry(channel)
+            .or_default()
+            .push(operation);
+    }
+}
+
+impl InternalChannel {
+    fn send(
+        &mut self,
+        fibers: &mut HashMap<FiberId, FiberTree>,
+        performing_fiber: Option<FiberId>,
+        packet: Packet,
+    ) {
+        self.pending_sends.push_back((performing_fiber, packet));
+        self.work_on_pending_operations(fibers);
+    }
+
+    fn receive(
+        &mut self,
+        fibers: &mut HashMap<FiberId, FiberTree>,
+        performing_fiber: Option<FiberId>,
+    ) {
+        self.pending_receives.push_back(performing_fiber);
+        self.work_on_pending_operations(fibers);
+    }
+
+    fn work_on_pending_operations(&mut self, fibers: &mut HashMap<FiberId, FiberTree>) {
+        if self.buffer.capacity == 0 {
+            while !self.pending_sends.is_empty() && !self.pending_receives.is_empty() {
+                let (send_id, packet) = self.pending_sends.pop_front().unwrap();
+                let receive_id = self.pending_receives.pop_front().unwrap();
+                Self::complete_send(fibers, send_id);
+                Self::complete_receive(fibers, receive_id, packet);
+            }
+        } else {
+            loop {
+                let mut did_perform_operation = false;
+
+                if !self.buffer.is_full() && let Some((fiber, packet)) = self.pending_sends.pop_front() {
+                    self.buffer.send(packet);
+                    Self::complete_send(fibers, fiber);
+                    did_perform_operation = true;
+                }
+
+                if !self.buffer.is_empty() && let Some(fiber) = self.pending_receives.pop_front() {
+                    let packet = self.buffer.receive();
+                    Self::complete_receive(fibers, fiber, packet);
+                    did_perform_operation = true;
+                }
+
+                if !did_perform_operation {
+                    break;
+                }
+            }
+        }
+    }
+
+    fn complete_send(fibers: &mut HashMap<FiberId, FiberTree>, fiber: Option<FiberId>) {
+        if let Some(fiber) = fiber {
+            let fiber = fibers.get_mut(&fiber).unwrap().as_single_mut().unwrap();
+            fiber.fiber.complete_send();
+        }
+    }
+    fn complete_receive(
+        fibers: &mut HashMap<FiberId, FiberTree>,
+        fiber: Option<FiberId>,
+        packet: Packet,
+    ) {
+        if let Some(fiber) = fiber {
+            let fiber = fibers.get_mut(&fiber).unwrap().as_single_mut().unwrap();
+            fiber.fiber.complete_receive(packet);
+        }
+    }
```

Comment:
Assuming we don't find a way to eliminate using `Option<FiberId>`, this would mean that all 6 call sites of `complete_send` and `complete_receive` would have an `if let` around the call.
I know that these functions are a weak abstraction layer – they only came to be after realizing that the code for performing a send or receive is the same 4 lines in several places. If you think we should really put the `if let` outside the function, I'd rather inline the whole function at the call sites.

Reply:
I also added a comment below at the top-level-thread to discuss `Option<FiberId>` in general.